File: leetcode3/최원준/449. Serialize and Deserialize BST.py
#

'''
1. 아이디어 :
pre_order로 순회하면서 str를 만듭니다.
pre_order는 가장 첫번째 숫자가 root, 그리고 왼쪽과 오른쪽을 나누는 경계선은 root의 숫자보다 커지기 시작하는 지점이므로,
binary search를 통해 계산합니다.

2. 시간복잡도 :
    O(n)
    O(nlogn)

3. 자료구조/알고리즘 :
pre_order traversal, binary search

'''

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

"""
     4
  2     6
1  3  5   7
"""

class Codec:

    def serialize(self, root: TreeNode | None) -> str:
        self.ans = ""

        def pre_order(node: TreeNode):
            if not node:
                return

            self.ans += str(node.val) + " "
            pre_order(node.left)
            pre_order(node.right)
        
        pre_order(root)
        return self.ans

        """Encodes a tree to a single string.
        """
        

    def deserialize(self, data: str) -> TreeNode | None:
        if not data:
            return None
        
        def build(lower, upper):
            if self.index == len(self.pre_order):
                return None
            
            val = self.pre_order[self.index]

            if val < lower or upper < val:
                return None
            
            self.index += 1
            par = TreeNode(val)
            par.left = build(lower, val)
            par.right = build(val, upper)

            return par

        self.pre_order = [int(d) for d in data.rstrip().split(" ")]
        self.index = 0
        return build(-float('inf'), float('inf'))
        """Decodes your encoded data to tree.
        """
        

# Your Codec object will be instantiated and called as such:
# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# tree = ser.serialize(root)
# ans = deser.deserialize(tree)
# return ans

File: leetcode3/이진희/449_Serialize and Deserialize BST.java
/*

1. 아이디어 : 이진탐색트리구조 규칙 활용
              직렬화: 선위순회로 풀이 -> StringBuilder를 사용하여 ","로 숫자를 구분하여 저장
              역질렬화: 문자열을 "," 기준으로 파싱 후, Dqeue를 사용해 TreeNode 복원
                        이때 노드 왼쪽 자식 값 < 노드 값 < 노드 오른쪽 자식 값의 규칙이 있는걸 명심

2. 시간복잡도 : 직렬화O(N) + 역직렬화O(N) => O(N)

3. 자료구조/알고리즘 : 이진탐색트리구조, Deque

 */

public class Codec {
    // 트리구조 -> 문자열
    public String serialize(TreeNode root) {
        if(root == null) return "";

        StringBuilder sb = new StringBuilder();

        serializeHelper(root, sb);
        return sb.toString();
    }

    private void serializeHelper(TreeNode root, StringBuilder sb) {
        if(root == null) return;

        sb.append(root.val).append(",");
        serializeHelper(root.left, sb);
        serializeHelper(root.right, sb);
    }

    // 문자열 -> 트리구조
    public TreeNode deserialize(String data) {
        if(data.isEmpty()) return null;

        String[] word = data.split(",");

        Deque<Integer> dq = new ArrayDeque<>();
        for(int i=0; i<word.length; i++) {
            dq.add(Integer.parseInt(word[i]));
        }

        return deserializeHelper(dq, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }

    private TreeNode deserializeHelper(Deque<Integer> dq, int min, int max) {
        if(dq.isEmpty()) return null;

        int val = dq.peek();
        if(val<min || val>max) return null;

        dq.poll();
        TreeNode root = new TreeNode(val);

        root.left = deserializeHelper(dq, min, val);
        root.right = deserializeHelper(dq, val, max);

        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// String tree = ser.serialize(root);
// TreeNode ans = deser.deserialize(tree);
// return ans;